File: TrappingRainWater.java
import java.util.*;
import java.io.*;

/*
Trapping Rain Water
Given an array arr[] of N non-negative integers representing height of blocks at index i as Ai where the width of each block is 1. Compute how much water can be trapped in between blocks after raining.
Structure is like below:
|  |
|_|
We can trap 2 units of water in the middle gap.



Input:
The first line of input contains an integer T denoting the number of test cases. The description of T test cases follows. Each test case contains an integer N denoting the size of the array, followed by N space separated numbers to be stored in array.

Output:
Output the total unit of water trapped in between the blocks.

Constraints:
1 <= T <= 100
3 <= N <= 107
0 <= Ai <= 108

Example:
Input:
2
4
7 4 0 9
3
6 9 9

Output:
10
0

Explanation:
Testcase 1: Water trapped by block of height 4 is 3 units, block of height 0 is 7 units. So, total unit of water trapped is 10 units.
*/

class TrappingRainWater
{
    static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
    public static void main (String[] args) throws Exception
	{
	    int T = Integer.parseInt(bf.readLine().trim());
	    for(int t=0; t<T; t++)
	    {
	        int N = Integer.parseInt(bf.readLine().trim());
	        String str[] = bf.readLine().split(" ");
	        
	        int arr[] = new int[N];
	        for(int i=0; i<N; i++)
	        {
	            arr[i] = Integer.parseInt(str[i]);
            }
            int result = 0; 
        
        //method 1   
            // maximum element on left and right 
            int left_max = 0, right_max = 0; 
               
            // indices to traverse the array 
            int lo = 0, hi = N-1; 
               
            while(lo <= hi)  
            { 
                if(arr[lo] < arr[hi]) 
                { 
                    if(arr[lo] > left_max)           
                        left_max = arr[lo]; // update max in left 
                    else          
                        result += left_max - arr[lo];                     
                    lo++; 
                } 
                else
                { 
                    if(arr[hi] > right_max)                       
                        right_max = arr[hi]; // update right maximum                       
                    else
                        result += right_max - arr[hi]; 
                    hi--; 
                } 
            } 
               
            System.out.println(result);
	    }
	}
}

File: JS/twoPointer/trappingRainWater.js
/*
https://leetcode.com/problems/trapping-rain-water/

42. Trapping Rain Water
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.

 

Example 1:
Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6
Explanation: The above elevation map (black section) is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped.

Example 2:
Input: height = [4,2,0,3,2,5]
Output: 9
*/


//approach 1 time: O(n), space: O(n)
var trap = function (height) {
    let n = height.length;

    let maxFromFront = [height[0]];
    let maxFromBack = [];

    let maxTillThis = height[0];
    for (let i = 1; i < n; i++) {
        maxTillThis = Math.max(maxTillThis, height[i]);
        maxFromFront[i] = maxTillThis;
    }

    maxTillThis = 0;
    for (let i = n - 1; i >= 0; i--) {
        maxTillThis = Math.max(maxTillThis, height[i]);
        maxFromBack[i] = maxTillThis;
    }

    let water = 0;
    for (let i = 1; i < n; i++) {
        water += (Math.min(maxFromFront[i], maxFromBack[i]) - height[i])
    }

    return water
};


//approach 2 - 2 pointer approach - time: O(n), space: O(1)
var trap = function (arr) {
    let n = arr.length;

    let left = 0, right = n - 1;
    let leftMax = 0, rightMax = 0;

    let water = 0;
    while (left <= right) {
        if (arr[left] < arr[right]) {
            if (arr[left] > leftMax) {
                leftMax = arr[left]
            } else {
                water += (leftMax - arr[left]);
            }
            left++
        } else {
            if (arr[right] > rightMax) {
                rightMax = arr[right]
            } else {
                water += (rightMax - arr[right]);
            }
            right--
        }
    }

    return water
};